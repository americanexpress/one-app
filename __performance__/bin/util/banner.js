/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const table = require('markdown-table');
const { Table } = require('console-table-printer');
const {
  testType, time, timeDiff, colors,
} = require('./format');
const { sum } = require('./math');

const metrics = {
  ...require('../k6/metrics'),
  ...require('../prometheus/metrics'),
};

function labelledMetric(metric, data, processor) {
  const { bold } = colors();
  return `${metrics[metric].label}: ${bold(metrics[metric].format(processor(data[metric])))}`;
}

function banner(colData, markdown = false) {
  const rows = [
    colData.map(({ meta }) => testType(meta.type)),
    colData.map(({ meta }) => time(meta.timestamp, 'Completed:')),
    colData.map(({ meta }) => timeDiff(meta.start, meta.end, 'Duration')),
    colData.map(({ data }) => labelledMetric('http_reqs', data, sum)),
    colData.map(({ data }) => labelledMetric('data_received', data, sum)),
    colData.map(({ data }) => labelledMetric('http_req_failed', data, sum)),
  ];

  if (colData.some(({ id }) => !!id)) rows.push(colData.map(({ id }) => id));

  if (markdown) {
    return table([
      colData.map(({ label }) => label),
      ...rows,
    ]);
  }

  const consoleTable = new Table({
    columns: colData.map(({ label }) => ({ name: label, alignment: 'left', color: 'none' })),
    colorMap: { none: '' },
  });

  consoleTable.addRows(rows.map((row) => row.reduce((acc, curr, i) => ({
    ...acc, [colData[i].label]: curr,
  }), {})));

  return consoleTable.render();
}

module.exports = banner;
