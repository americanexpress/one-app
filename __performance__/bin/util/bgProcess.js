/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const fs = require('node:fs');
const { join } = require('node:path');
const { tmpdir } = require('node:os');
const pkg = require('../../../package.json');

const dir = join(tmpdir(), `one-perf-${pkg.version}`);
const filePath = join(dir, 'processes.json');

fs.mkdirSync(dir, { recursive: true });

const getLogFile = (name) => join(dir, `${name}.log`);

async function saveProcess(name, pid) {
  let processes = {};
  if (fs.existsSync(filePath)) {
    try {
      processes = JSON.parse(fs.promises.readFile(filePath));
    } catch (error) { /* ignore */ }
  }
  processes[name] = pid;
  await fs.promises.writeFile(filePath, JSON.stringify(processes, null, 2));
}

async function killProcess(name) {
  if (!fs.existsSync(filePath)) {
    console.error('No processes found.');
    process.exit(1);
  }
  const processes = JSON.parse(await fs.promises.readFile(filePath));
  const pid = processes[name];
  if (!pid) {
    console.error(`No process found with the name "${name}".`);
    process.exit(1);
  }
  console.log(`Killing process ${name}:${pid}...`);
  try {
    process.kill(pid, 'SIGTERM');
  } catch (error) {
    console.error(`Failed to kill process ${name}:${pid}. Continuing cleanup...`);
  }
  const logFile = getLogFile(name);
  await fs.promises.rm(logFile);
  delete processes[name];
  if (Object.keys(processes).length === 0) {
    await fs.promises.rm(filePath);
  } else {
    await fs.promises.writeFile(filePath, JSON.stringify(processes, null, 2));
  }
  process.exit(0);
}

async function retrieveProcess(name) {
  if (!fs.existsSync(filePath)) {
    console.error('No processes found.');
    process.exit(1);
  }
  const processes = JSON.parse(await fs.promises.readFile(filePath));
  const pid = processes[name];
  if (!pid) {
    console.error(`No process found with the name "${name}".`);
    process.exit(1);
  }
  return {
    name,
    pid,
    logFile: getLogFile(name),
  };
}

module.exports = {
  getLogFile,
  retrieveProcess,
  killProcess,
  saveProcess,
};
