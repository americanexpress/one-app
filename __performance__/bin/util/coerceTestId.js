/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const fs = require('node:fs');
const path = require('node:path');
const { diffArrays } = require('diff');
const {
  green, red, dim, underline, strikethrough,
} = require('colorette');
const getResultDir = require('./getResultDir');

function createCoerceTestId(optionName) {
  return function coerceTestId(arg) {
    const dir = /^[./~]/.test(arg) ? path.resolve(process.cwd(), arg) : getResultDir(arg);
    const id = path.basename(dir);

    const dirContents = fs.readdirSync(dir).sort();
    const expectedContents = ['k6-out.json', 'k6-summary.json', 'metrics-k6.json', 'metrics-prom.json'].sort();

    if (!expectedContents.every((file) => dirContents.includes(file))) {
      const contentsDiff = diffArrays(expectedContents, dirContents, { ignoreWhitespace: true });
      const printDiff = contentsDiff.map((part) => (part.added ? strikethrough(green(part.value))
        // eslint-disable-next-line unicorn/no-nested-ternary -- conflicting rules
        : part.removed ? red(part.value) : dim(part.value)));

      throw new Error(`${red(`Bad option ${underline(optionName)}:`)}\n  Invalid test result directory: ${dir}\n\n  ${printDiff.join(', ')}`);
    }

    return { dir, id };
  };
}

module.exports = createCoerceTestId;
