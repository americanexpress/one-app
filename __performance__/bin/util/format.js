/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const { formatRelative, formatDistanceStrict } = require('date-fns');
const { createColors } = require('colorette');

function colors(supportMarkdown = true) {
  const useMarkdown = process.env.MARKDOWN === 'true';
  const colorette = createColors({ useColor: !useMarkdown });
  if (!useMarkdown || !supportMarkdown) return colorette;
  return {
    ...colorette,
    bold: (str) => `**${str}**`,
    italic: (str) => `_${str}_`,
    strikethrough: (str) => `~~${str}~~`,
    underline: (str) => `<u>${str}</u>`,
  };
}

function nth(n) {
  switch (true) {
    case n % 100 === 11:
    case n % 100 === 12:
    case n % 100 === 13:
      return `${n}th`;
    case n % 10 === 1:
      return `${n}st`;
    case n % 10 === 2:
      return `${n}nd`;
    case n % 10 === 3:
      return `${n}rd`;
    default:
      return `${n}th`;
  }
}

function float(n, precision = 3) {
  return Number(n.toFixed(precision)).toString();
}

function append(unit, format = float) {
  return (n) => `${format(n)}${unit}`;
}

function percent(n) {
  return `${float(n * 100)}%`;
}

function bytes(n) {
  if (n === 0) return '0B';
  if (n < 1e3) return `${float(n, 0)}B`;
  if (n < 1e6) return `${float(n / 1e3)}kB`;
  if (n < 1e9) return `${float(n / 1e6)}MB`;
  return `${float(n / 1e9)}GB`;
}

function pValue(p, alpha = 0.05) {
  const { red, bold } = colors();
  if (p < 0.001) return red(bold('*<0.001'));
  if (p <= alpha) return red(`*${float(p)}`);
  return float(p);
}

function seconds(n) {
  if (n === 0) return '0s';
  if (n < 1e-6) return `${float(n * 1e9)}ns`;
  if (n < 1e-3) return `${float(n * 1e6)}Âµs`;
  if (n < 1) return `${float(n * 1e3)}ms`;
  if (n < 60) return `${float(n, 2)}s`;
  if (n < 3600) return `${Math.floor(n / 60)}m ${seconds(n % 60)}`;
  return `${Math.floor(n / 3600)}h ${seconds(n % 3600)}`;
}

function milliseconds(n) {
  return seconds(n / 1e3);
}

function diff(c, v, format) {
  const { green, red, bold } = colors(false);
  const d = c - v;
  if (d === 0) return `=${format(d)}`;
  if (d > 0) return green(`${bold('-')}${format(d)} (${bold('-')}${percent(d / c)})`);
  return red(`${bold('+')}${format(d * -1)} (${bold('+')}${percent((d / c) * -1)})`);
}

function capitalize(string) {
  return string.replace(/^\w/, (c) => c.toUpperCase());
}

function testType(type) {
  const { bold } = colors();
  return `Test type: ${bold(capitalize(type))}`;
}

function time(timestamp, prefix) {
  const { bold } = colors();
  const relative = formatRelative(timestamp, Date.now());
  if (prefix) return `${prefix} ${bold(capitalize(relative))}`;
  return capitalize(relative);
}

function timeDiff(start, end, prefix) {
  const { bold } = colors();
  const formatted = formatDistanceStrict(end, start);
  if (prefix) return `${prefix} ${bold(formatted)}`;
  return formatted;
}

module.exports = {
  nth,
  colors,
  diff,
  float,
  append,
  percent,
  bytes,
  pValue,
  seconds,
  milliseconds,
  capitalize,
  testType,
  time,
  timeDiff,
};
