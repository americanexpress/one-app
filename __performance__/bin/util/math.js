/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

function sum(data) {
  return data.reduce((acc, value) => acc + value, 0);
}
sum.label = 'Total';

function mean(data) {
  return sum(data) / data.length;
}
mean.label = 'Average';

function median(data) {
  const sorted = data.sort((a, b) => a - b);
  return (sorted[Math.floor((data.length - 1) / 2)] + sorted[Math.ceil((data.length - 1) / 2)]) / 2;
}
median.label = 'Median';

function mode(data) {
  const frequencyMap = {};
  let maxFreq = 0;
  let result;
  for (const item of data) {
    frequencyMap[item] = ++frequencyMap[item] || 1;
    if (frequencyMap[item] > maxFreq) {
      maxFreq = frequencyMap[item];
      result = item;
    }
  }
  return result;
}
mode.label = 'Mode';

function min(data) {
  return Math.min(...data);
}
min.label = 'Min';

function max(data) {
  return Math.max(...data);
}
max.label = 'Max';

function percentile(p) {
  return (data) => data.sort((a, b) => a - b)[Math.floor(data.length * (p / 100)) - 1];
}

const p95 = percentile(95);
p95.label = '95th Percentile';

const p90 = percentile(90);
p90.label = '90th Percentile';

module.exports = {
  sum,
  mean,
  median,
  mode,
  min,
  max,
  percentile,
  p95,
  p90,
};
