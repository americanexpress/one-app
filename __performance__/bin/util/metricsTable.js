/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const table = require('markdown-table');
const { Table } = require('console-table-printer');
const k6metrics = require('../k6/metrics');
const promMetrics = require('../prometheus/metrics');

const metrics = { ...k6metrics, ...promMetrics };

function generateMetricsTable({
  data,
  headers,
  processors,
  markdown = false,
  description = false,
  raw = false,
}) {
  const rows = Object.keys(data).map((metric) => [
    metrics[metric].label,
    description && (metrics[metric].description || ' '),
    ...processors.map((processor) => {
      if (processor.length === 1) {
        return raw ? processor(data[metric]) : metrics[metric].format(processor(data[metric]));
      }
      return processor(metric, data[metric], raw ? (x) => x : metrics[metric].format);
    }),
  ].filter(Boolean));

  const tableHeaders = ['Metric', description && 'Description', ...headers].filter(Boolean);

  if (markdown) {
    return table([
      tableHeaders,
      ...rows,
    ]);
  }

  const consoleTable = new Table({
    columns: [
      { name: 'Metric', alignment: 'left', color: 'cyan' },
      description && {
        name: 'Description', alignment: 'left', maxLen: 40, color: 'none',
      },
      ...headers.map((name) => ({ name, alignment: 'left', color: 'none' })),
    ].filter(Boolean),
    colorMap: { none: '' },
  });

  consoleTable.addRows(rows.map((row) => row.reduce((acc, col, i) => ({
    ...acc,
    [tableHeaders[i]]: col,
  }), {})));

  return consoleTable.render();
}

module.exports = generateMetricsTable;
