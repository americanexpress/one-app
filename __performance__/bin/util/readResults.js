/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const fs = require('node:fs/promises');
const path = require('node:path');

async function readResults(resultsDir) {
  const [k6, prom, summary] = (await Promise.all([
    fs.readFile(path.join(resultsDir, 'metrics-k6.json')),
    fs.readFile(path.join(resultsDir, 'metrics-prom.json')),
    fs.readFile(path.join(resultsDir, 'k6-summary.json')),
  ])).map(JSON.parse);

  return {
    data: { ...k6, ...prom },
    summary,
    get meta() {
      return summary.meta;
    },
  };
}

module.exports = readResults;
