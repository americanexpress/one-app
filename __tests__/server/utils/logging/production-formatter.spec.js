/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import util from 'util';

jest.mock('../../../../src/server/utils/readJsonFile', () => () => ({ buildVersion: '1.2.3-abc123' }));
jest.mock('os', () => ({
  hostname: jest.fn(() => 'host-123'),
  type: jest.fn(() => 'Darwin'),
  arch: jest.fn(() => 'x64'),
}));

function createErrorWithMockedStacktrace(msg) {
  const err = new Error(msg);
  err.stack = `Error: test error
  at Object.<anonymous> (one-app/__tests__/server/utils/logging/production-formatter.spec.js:36:64)
  at Object.asyncFn (one-app/node_modules/jest-jasmine2/build/jasmine_async.js:124:345)
  at resolve (one-app/node_modules/jest-jasmine2/build/queue_runner.js:46:12)
  at new Promise (<anonymous>)
  at mapper (one-app/node_modules/jest-jasmine2/build/queue_runner.js:34:499)
  at promise.then (one-app/node_modules/jest-jasmine2/build/queue_runner.js:74:39)
  at <anonymous>
  at process._tickCallback (internal/process/next_tick.js:188:7)`;

  return err;
}

describe('production-formatter', () => {
  let formatter;
  jest.spyOn(Date.prototype, 'toISOString').mockImplementation(() => '2018-03-02T02:39:32.948Z');
  jest.spyOn(util, 'format');

  function load(pid) {
    jest.resetModules();

    Object.defineProperty(process, 'pid', {
      writable: true,
      value: pid || 1234,
    });
    formatter = require('../../../../src/server/utils/logging/production-formatter').default;
  }

  it('is a function', () => {
    load();
    expect(typeof formatter).toBe('function');
  });

  it('encodes as parseable JSON', () => {
    load();
    const entry = formatter('error', 1, 2, 3);
    expect(() => JSON.parse(entry)).not.toThrowError();
    expect(JSON.parse(entry)).toMatchSnapshot();
  });

  it('drops lower levels', () => {
    load();
    expect(formatter('debug', 1, 2, 3)).toBe(null);
  });

  describe('errors', () => {
    it('encodes as parseable JSON', () => {
      load();
      const entry = formatter('error', 'unable to do the thing', createErrorWithMockedStacktrace('test error'));
      expect(() => JSON.parse(entry)).not.toThrowError();
      expect(JSON.parse(entry)).toMatchSnapshot();
    });

    it('puts a lone error in the error field', () => {
      expect.assertions(2);
      load();
      const parsed = JSON.parse(formatter('error', createErrorWithMockedStacktrace('test error')));
      expect(parsed).toHaveProperty('error');
      expect(parsed.error).toMatchSnapshot();
    });

    it('does not include a message key when given a lone error', () => {
      expect.assertions(1);
      load();
      const parsed = JSON.parse(formatter('error', new Error('irrelevant')));
      expect(parsed).not.toHaveProperty('message');
    });

    it('puts an error with other arguments in the error field and uses only the other args for the message', () => {
      expect.assertions(4);
      load();
      const parsed = JSON.parse(formatter('error', createErrorWithMockedStacktrace('test error'), 'say hi %i times', 7));
      expect(parsed).toHaveProperty('error');
      expect(parsed).toHaveProperty('message');
      expect(parsed.error).toMatchSnapshot();
      expect(parsed.message).toBe('say hi 7 times');
    });

    it('puts an error paired with a message in the error field', () => {
      expect.assertions(4);
      load();
      const parsed = JSON.parse(formatter('error', 'unable to do the thing', createErrorWithMockedStacktrace('test error')));
      expect(parsed).toHaveProperty('error');
      expect(parsed).toHaveProperty('message');
      expect(parsed.error).toMatchSnapshot();
      expect(parsed.message).toBe('unable to do the thing');
    });

    it('uses the error in the message when the error is paired with a message and other arguments', () => {
      expect.assertions(4);
      load();
      const parsed = JSON.parse(formatter('error', 'unable to do the thing %O %d times', createErrorWithMockedStacktrace('test error'), 12));
      const errorText = parsed.message
        .replace('[', '')
        .replace(']', '');
      expect(parsed).toHaveProperty('error');
      expect(parsed).toHaveProperty('message');
      expect(parsed.error).toMatchSnapshot();
      expect(errorText).toMatchSnapshot();
    });

    it('records the error message as <none> if not present', () => {
      expect.assertions(2);
      load();
      const err = createErrorWithMockedStacktrace('test error');
      delete err.message;
      const parsed = JSON.parse(formatter('error', 'unable to do the thing', err));
      expect(parsed).toHaveProperty('error');
      expect(parsed.error).toHaveProperty('message', '<none>');
    });

    it('records the error stacktrace as <none> if not present', () => {
      expect.assertions(2);
      load();
      const err = createErrorWithMockedStacktrace('test error');
      delete err.stack;
      const parsed = JSON.parse(formatter('error', 'unable to do the thing', err));
      expect(parsed).toHaveProperty('error');
      expect(parsed.error).toHaveProperty('stacktrace', '<none>');
    });

    it('encodes ClientReportedError Error as parseable JSON', () => {
      function buildClientErrorEntry() {
        const error = new Error('something broke');
        Object.assign(error, {
          name: 'ClientReportedError',
          stack: 'Error: something broke\n    at methodA (resource-a.js:1:1)\n    at methodB (resource-b.js:1:1)\n',
          userAgent: 'Browser/5.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)',
          uri: 'https://example.com/page-the/error/occurred-on',
          metaData: {
            moduleID: 'healthy-frank',
            code: '500',
            collectionMethod: 'applicationError',
            correlationId: '123abc',
          },
        });
        return error;
      }

      load();
      const entry = formatter('error', buildClientErrorEntry());
      expect(() => JSON.parse(entry)).not.toThrowError();
      expect(JSON.parse(entry)).toMatchSnapshot();
    });

    it('encodes Server Reported Error as parseable JSON with nested metadata objects', () => {
      function buildServerSideErrorEntry() {
        const error = new Error('something broke');
        Object.assign(error, {
          stack: 'Error: something broke\n    at methodA (resource-a.js:1:1)\n    at methodB (resource-b.js:1:1)\n',
          metaData: {
            moduleID: 'healthy-frank',
            nestedObject: {
              level1: {
                level2: {
                  level3: {
                    level4: 'logs nested objects correctly',
                  },
                },
              },
            },
          },
        });
        return error;
      }

      load();
      const entry = formatter('error', buildServerSideErrorEntry());
      expect(() => JSON.parse(entry)).not.toThrowError();
      expect(JSON.parse(entry)).toMatchSnapshot();
    });
  });

  describe('types', () => {
    it('encodes request as parseable JSON', () => {
      function buildRequestEntry() {
        return {
          type: 'request',
          request: {
            direction: 'in',
            protocol: 'https',
            address: {
              url: 'https://example.org/server',
            },
            metaData: {
              method: 'GET',
              correlationId: '123',
              // null to explicitly signal no value, undefined if not expected for every request
              host: 'example.org',
              referrer: 'https://example.org/other-page',
              userAgent: 'Browser/8.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)',
              location: undefined,
            },
            timings: { // https://www.w3.org/TR/navigation-timing/
              // navigationStart: 0,
              // redirectStart: 0,
              // redirectEnd: 0,
              // fetchStart: 0,
              // domainLookupStart: 10,
              // domainLookupEnd: 20,
              // connectStart: 30,
              // secureConnectionStart: 40,
              // connectEnd: 50,
              // requestStart: 60,
              // requestEnd: 70, // optional? not part of the W3C spec
              // responseStart: 80,
              // responseEnd: 90,
              duration: 10,
              ttfb: 20,
            },
            statusCode: 200,
            statusText: 'OK',
          },
        };
      }

      load();
      const entry = formatter('error', buildRequestEntry());
      expect(() => JSON.parse(entry)).not.toThrowError();
      expect(JSON.parse(entry)).toMatchSnapshot();
    });

    it('uses the default util.format when given an unknown type', () => {
      load();
      util.format.mockClear();
      const entry = formatter('error', { type: 'yolo' });
      expect(util.format).toHaveBeenCalledTimes(1);
      expect(() => JSON.parse(entry)).not.toThrowError();
      expect(JSON.parse(entry)).toMatchSnapshot();
    });
  });
});
