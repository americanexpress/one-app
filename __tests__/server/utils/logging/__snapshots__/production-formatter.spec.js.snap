// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`production-formatter encodes as parseable JSON 1`] = `
Object {
  "application": Object {
    "name": "One App",
    "version": "1.2.3-abc123",
  },
  "device": Object {
    "agent": "Darwin x64",
    "id": "host-123:1234",
  },
  "level": "error",
  "message": "1 2 3",
  "schemaVersion": "0.3.0",
  "timestamp": "2018-03-02T02:39:32.948Z",
}
`;

exports[`production-formatter errors encodes ClientReportedError Error as parseable JSON 1`] = `
Object {
  "application": Object {
    "name": "One App",
    "version": "1.2.3-abc123",
  },
  "device": Object {
    "agent": "Browser/5.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)",
  },
  "error": Object {
    "message": "something broke",
    "name": "ClientReportedError",
    "stacktrace": "Error: something broke
    at methodA (resource-a.js:1:1)
    at methodB (resource-b.js:1:1)
",
  },
  "level": "error",
  "request": Object {
    "address": Object {
      "uri": "https://example.com/page-the/error/occurred-on",
    },
    "metaData": Object {
      "code": "500",
      "collectionMethod": "applicationError",
      "correlationId": "123abc",
      "moduleID": "healthy-frank",
    },
  },
  "schemaVersion": "0.3.0",
  "timestamp": "2018-03-02T02:39:32.948Z",
}
`;

exports[`production-formatter errors encodes Server Rerported Error as parseable JSON with nested metadata objects 1`] = `
Object {
  "application": Object {
    "name": "One App",
    "version": "1.2.3-abc123",
  },
  "device": Object {
    "agent": "Darwin x64",
    "id": "host-123:1234",
  },
  "error": Object {
    "message": "something broke",
    "name": "Error",
    "stacktrace": "Error: something broke
    at methodA (resource-a.js:1:1)
    at methodB (resource-b.js:1:1)
",
  },
  "level": "error",
  "metaData": Object {
    "moduleID": "healthy-frank",
    "nestedObject": Object {
      "level1": Object {
        "level2": Object {
          "level3": Object {
            "level4": "logs nested objects correctly",
          },
        },
      },
    },
  },
  "schemaVersion": "0.3.0",
  "timestamp": "2018-03-02T02:39:32.948Z",
}
`;

exports[`production-formatter errors encodes as parseable JSON 1`] = `
Object {
  "application": Object {
    "name": "One App",
    "version": "1.2.3-abc123",
  },
  "device": Object {
    "agent": "Darwin x64",
    "id": "host-123:1234",
  },
  "error": Object {
    "message": "test error",
    "name": "Error",
    "stacktrace": "Error: test error
  at Object.<anonymous> (one-app/__tests__/server/utils/logging/production-formatter.spec.js:36:64)
  at Object.asyncFn (one-app/node_modules/jest-jasmine2/build/jasmine_async.js:124:345)
  at resolve (one-app/node_modules/jest-jasmine2/build/queue_runner.js:46:12)
  at new Promise (<anonymous>)
  at mapper (one-app/node_modules/jest-jasmine2/build/queue_runner.js:34:499)
  at promise.then (one-app/node_modules/jest-jasmine2/build/queue_runner.js:74:39)
  at <anonymous>
  at process._tickCallback (internal/process/next_tick.js:188:7)",
  },
  "level": "error",
  "message": "unable to do the thing",
  "schemaVersion": "0.3.0",
  "timestamp": "2018-03-02T02:39:32.948Z",
}
`;

exports[`production-formatter errors puts a lone error in the error field 1`] = `
Object {
  "message": "test error",
  "name": "Error",
  "stacktrace": "Error: test error
  at Object.<anonymous> (one-app/__tests__/server/utils/logging/production-formatter.spec.js:36:64)
  at Object.asyncFn (one-app/node_modules/jest-jasmine2/build/jasmine_async.js:124:345)
  at resolve (one-app/node_modules/jest-jasmine2/build/queue_runner.js:46:12)
  at new Promise (<anonymous>)
  at mapper (one-app/node_modules/jest-jasmine2/build/queue_runner.js:34:499)
  at promise.then (one-app/node_modules/jest-jasmine2/build/queue_runner.js:74:39)
  at <anonymous>
  at process._tickCallback (internal/process/next_tick.js:188:7)",
}
`;

exports[`production-formatter errors puts an error paired with a message in the error field 1`] = `
Object {
  "message": "test error",
  "name": "Error",
  "stacktrace": "Error: test error
  at Object.<anonymous> (one-app/__tests__/server/utils/logging/production-formatter.spec.js:36:64)
  at Object.asyncFn (one-app/node_modules/jest-jasmine2/build/jasmine_async.js:124:345)
  at resolve (one-app/node_modules/jest-jasmine2/build/queue_runner.js:46:12)
  at new Promise (<anonymous>)
  at mapper (one-app/node_modules/jest-jasmine2/build/queue_runner.js:34:499)
  at promise.then (one-app/node_modules/jest-jasmine2/build/queue_runner.js:74:39)
  at <anonymous>
  at process._tickCallback (internal/process/next_tick.js:188:7)",
}
`;

exports[`production-formatter errors puts an error with other arguments in the error field and uses only the other args for the message 1`] = `
Object {
  "message": "test error",
  "name": "Error",
  "stacktrace": "Error: test error
  at Object.<anonymous> (one-app/__tests__/server/utils/logging/production-formatter.spec.js:36:64)
  at Object.asyncFn (one-app/node_modules/jest-jasmine2/build/jasmine_async.js:124:345)
  at resolve (one-app/node_modules/jest-jasmine2/build/queue_runner.js:46:12)
  at new Promise (<anonymous>)
  at mapper (one-app/node_modules/jest-jasmine2/build/queue_runner.js:34:499)
  at promise.then (one-app/node_modules/jest-jasmine2/build/queue_runner.js:74:39)
  at <anonymous>
  at process._tickCallback (internal/process/next_tick.js:188:7)",
}
`;

exports[`production-formatter errors uses the error in the message when the error is paired with a message and other arguments 1`] = `
Object {
  "message": "test error",
  "name": "Error",
  "stacktrace": "Error: test error
  at Object.<anonymous> (one-app/__tests__/server/utils/logging/production-formatter.spec.js:36:64)
  at Object.asyncFn (one-app/node_modules/jest-jasmine2/build/jasmine_async.js:124:345)
  at resolve (one-app/node_modules/jest-jasmine2/build/queue_runner.js:46:12)
  at new Promise (<anonymous>)
  at mapper (one-app/node_modules/jest-jasmine2/build/queue_runner.js:34:499)
  at promise.then (one-app/node_modules/jest-jasmine2/build/queue_runner.js:74:39)
  at <anonymous>
  at process._tickCallback (internal/process/next_tick.js:188:7)",
}
`;

exports[`production-formatter errors uses the error in the message when the error is paired with a message and other arguments 2`] = `
"unable to do the thing Error: test error
  at Object.<anonymous> (one-app/__tests__/server/utils/logging/production-formatter.spec.js:36:64)
  at Object.asyncFn (one-app/node_modules/jest-jasmine2/build/jasmine_async.js:124:345)
  at resolve (one-app/node_modules/jest-jasmine2/build/queue_runner.js:46:12)
  at new Promise (<anonymous>)
  at mapper (one-app/node_modules/jest-jasmine2/build/queue_runner.js:34:499)
  at promise.then (one-app/node_modules/jest-jasmine2/build/queue_runner.js:74:39)
  at <anonymous>
  at process._tickCallback (internal/process/next_tick.js:188:7) 12 times"
`;

exports[`production-formatter types encodes request as parseable JSON 1`] = `
Object {
  "application": Object {
    "name": "One App",
    "version": "1.2.3-abc123",
  },
  "device": Object {
    "agent": "Darwin x64",
    "id": "host-123:1234",
  },
  "level": "error",
  "request": Object {
    "address": Object {
      "url": "https://example.org/server",
    },
    "direction": "in",
    "metaData": Object {
      "correlationId": "123",
      "host": "example.org",
      "method": "GET",
      "referrer": "https://example.org/other-page",
      "userAgent": "Browser/8.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)",
    },
    "protocol": "https",
    "statusCode": 200,
    "statusText": "OK",
    "timings": Object {
      "duration": 10,
      "ttfb": 20,
    },
  },
  "schemaVersion": "0.3.0",
  "timestamp": "2018-03-02T02:39:32.948Z",
}
`;

exports[`production-formatter types uses the default util.format when given an unknown type 1`] = `
Object {
  "application": Object {
    "name": "One App",
    "version": "1.2.3-abc123",
  },
  "device": Object {
    "agent": "Darwin x64",
    "id": "host-123:1234",
  },
  "level": "error",
  "message": "{ type: 'yolo' }",
  "schemaVersion": "0.3.0",
  "timestamp": "2018-03-02T02:39:32.948Z",
}
`;
